File: scraper.js
const playwright = require('playwright');

let movieHash = []

function getMovieInfo(listItem) {
    let title = listItem.querySelector('h5 a') ? listItem.querySelector('h5 a').innerText.replaceAll('"', '') : ""
    let url = listItem.querySelector('a') ? listItem.querySelector('a').href : ""
    let category = listItem.querySelector('.badge-split-film-category .badge-content') ? listItem.querySelector('.badge-split-film-category .badge-content').innerText : ""
    let genres = listItem.querySelector('.badge-split-film-genre .badge-content') ? listItem.querySelector('.badge-split-film-genre .badge-content').innerText.replaceAll(',', ' | ') : ""
    let section = listItem.querySelector('.badge-split-film-section .badge-content') ? listItem.querySelector('.badge-split-film-section .badge-content').innerText.replaceAll(',', ' | ') : ""
    let premiere = listItem.querySelector('.badge-split-film-premiere .badge-content') ? listItem.querySelector('.badge-split-film-premiere .badge-content').innerText.replaceAll(',', ' | ') : ""
    let director = listItem.querySelector('.badge-split-director .badge-content') ? listItem.querySelector('.badge-split-director .badge-content').innerText.replaceAll(',', ' | ') : ""
    let image_url = listItem.querySelector('.er-event-image img').src ? listItem.querySelector('.er-event-image img').src : ""

    let hash = {
        'title': title,
        'url': url,
        'category': category,
        'genres': genres,
        'section': section,
        'premiere': premiere,
        'director': director,
        'image_url': image_url,
    };
    movieHash.push(hash);
}

function convertToJson(hash) {
    let json = hash
    let fields = Object.keys(json[0])
    let replacer = function (key, value) { return value === null ? '' : value }
    let csv = json.map(function (row) {
        return fields.map(function (fieldName) {
            return JSON.stringify(row[fieldName], replacer)
        }).join(',')
    })
    csv.unshift(fields.join(',')) // add header column
    csv = csv.join('\r\n');
    copy(csv)
    console.log(csv)
}

(async () => {
    const browser = await playwright.chromium.launch({ headless: true });
    const context = await browser.newContext();
    const page = await context.newPage();
    await page.goto("https://schedule.sxsw.com/2024/search/film"); // TODO: Use current year

    let movieNodes = page.querySelectorAll('.er-content-container')
    

    // populate movieHash
    movieNodes.forEach(node => getMovieInfo(node))

    convertToJson(movieHash)

console.log("this is wehre the magic happnes")
    await browser.close(); // TODO: close earlier?
    // return result
})();

File: scraper.py
import json
from bs4 import BeautifulSoup
from playwright.sync_api import sync_playwright

movie_hash = []

def get_movie_info(node):
        # Fetch the HTML content of the node
    node_html = node.inner_html()
    # Create a Beautiful Soup object for the node
    soup = BeautifulSoup(node_html, 'html.parser')

    title = soup.select_one('h5 a').text.replace('"', '') if soup.select_one('h5 a') else ""
    url = soup.select_one('a')['href'] if soup.select_one('a') else ""
    category = soup.select_one('.badge-split-film-category .badge-content').text if soup.select_one('.badge-split-film-category .badge-content') else ""
    genres = soup.select_one('.badge-split-film-genre .badge-content').text.replace(',', ' | ') if soup.select_one('.badge-split-film-genre .badge-content') else ""
    section = soup.select_one('.badge-split-film-section .badge-content').text.replace(',', ' | ') if soup.select_one('.badge-split-film-section .badge-content') else ""
    premiere = soup.select_one('.badge-split-film-premiere .badge-content').text.replace(',', ' | ') if soup.select_one('.badge-split-film-premiere .badge-content') else ""
    director = soup.select_one('.badge-split-director .badge-content').text.replace(',', ' | ') if soup.select_one('.badge-split-director .badge-content') else ""
    image_url = soup.select_one('.er-event-image img')['src'] if soup.select_one('.er-event-image img') else ""

    print("Cleaning up movie data for: " + title)

    hash = {
        'title': title,
        'url': url,
        'category': category,
        'genres': genres,
        'section': section,
        'premiere': premiere,
        'director': director,
        'image_url': image_url,
    }
    movie_hash.append(hash)

def convert_to_json(hash):
    json_output = json.dumps(hash, ensure_ascii=False)
    print(json_output)
    # For copying to clipboard, you might need additional libraries like pyperclip
    # pyperclip.copy(json_output)


with sync_playwright() as p:
    browser = p.chromium.launch()
    page = browser.new_page()

    print("Opening page...")

    page.goto("https://arelia.github.io/arelias-sxsw-film-data-grabber/")

    print("Grabbing movie nodes...")

    movie_nodes = page.query_selector_all('.er-content-container')

    print("Parsing movie nodes...")

    [get_movie_info(movie) for movie in movie_nodes]

    # print("Our movie_hash:", movie_hash)

    print("Converting to JSON...")

    convert_to_json(movie_hash)

    print("Done!")

    browser.close(); # TODO: close earlier?
